starlette/middleware/base.py

```diff
@@ -14,6 +14,68 @@
 T = typing.TypeVar("T")
 
 
+class _CachedRequest(Request):
+    """
+    If the user calls Request.body() from their dispatch function
+    we cache the entire request body in memory and pass that to downstream middlewares,
+    but if they call Request.stream() then all we do is send an
+    empty body so that downstream things don't hang forever.
+    """
+
+    def __init__(self, *args: typing.Any, **kwargs: typing.Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._wrapped_rcv_disconnected = False
+        self._wrapped_rcv_consumed = False
+
+    async def wrapped_receive(self) -> Message:
+        wrapped_rcv_connected = not (
+            self._wrapped_rcv_disconnected or self._wrapped_rcv_consumed
+        )
+        if wrapped_rcv_connected:
+            if getattr(self, "_body", None) is not None:
+                # body() was called, we return it even if the client disconnected
+                self._wrapped_rcv_consumed = True
+                return {
+                    "type": "http.request",
+                    "body": self._body,
+                    "more_body": False,
+                }
+            elif self._stream_consumed:
+                # stream() was called to completion or client disconnected
+                self._wrapped_rcv_consumed = True
+                return {
+                    "type": "http.request",
+                    "body": b"",
+                    "more_body": False,
+                }
+            else:
+                # body() was never called and stream() wasn't consumed
+                stream = self.stream()
+                try:
+                    chunk = await stream.__anext__()
+                    self._wrapped_rcv_consumed = self._stream_consumed
+                    return {
+                        "type": "http.request",
+                        "body": chunk,
+                        "more_body": self._stream_consumed,
+                    }
+                except ClientDisconnect:
+                    self._wrapped_rcv_disconnected = True
+                    return {"type": "http.disconnect"}
+        # wrapped_rcv is either disconnected or consumed
```

Comment:
The "or consumed" also shouldn't be here, right?

Reply:
I'll re-organize this a bit to clarify